File: runtime/src/test/py/list_page.py
#
# microtome

from microtome.page import Page
from microtome.prop.prop_spec import PropSpec
from microtome.prop.prop import Prop
from primitive_page import PrimitivePage
import types

class ListPage(Page):
    def __init__(self, name):
        super(ListPage, self).__init__(name)
        self._kids = Prop(self, ListPage._s_kidsSpec)

    @property
    def kids(self):
        return self._kids.value

    @property
    def props(self):
        return super(ListPage, self).props + [self._kids]

    _s_kidsSpec = PropSpec("kids", None, [types.ListType, PrimitivePage])

File: runtimes/objc/Tests/ListPage.m

#import "ListPage.h"
#import "PrimitivePage.h"
#import "microtome.h"

static MTPropSpec* s_kidsSpec = nil;

@implementation ListPage {
@protected
    MTObjectProp* _kids;
}

- (NSArray*)kids { return _kids.value; }

- (NSArray*)props { return MT_PROPS(_kids, ); }

- (id)init {
    if ((self = [super init])) {
        _kids = [[MTObjectProp alloc] initWithPropSpec:s_kidsSpec];
    }
    return self;
}

+ (void)initialize {
    if (self == [ListPage class]) {
        s_kidsSpec = [[MTPropSpec alloc] initWithName:@"kids"
            annotations:nil
            valueClasses:@[ [NSArray class],[PrimitivePage class], ] ];
    }
}

@end
